src/util/tpool/fd_tpool.h

```diff
@@ -0,0 +1,932 @@
+  FD_VOLATILE( worker->task_l0     ) = task_l0;     FD_VOLATILE( worker->task_l1     ) = task_l1;
+  FD_VOLATILE( worker->task_m0     ) = task_m0;     FD_VOLATILE( worker->task_m1     ) = task_m1;
+  FD_VOLATILE( worker->task_n0     ) = task_n0;     FD_VOLATILE( worker->task_n1     ) = task_n1;
+  FD_COMPILER_MFENCE();
+  FD_VOLATILE( worker->state ) = FD_TPOOL_WORKER_STATE_EXEC;
+  FD_COMPILER_MFENCE();
+}
+
+/* fd_tpool_waits waits for the tpool thread worker_idx to leave the
+   EXEC state.  As this is used in high performance contexts, does no
+   input argument checking.  Specifically, assumes tpool is valid,
+   worker_idx in (0,worker_cnt) (yes, open on both ends) and caller is
+   not tpool thread worker_idx.  worker_idx 0 is considered to always be
+   in an exec state so we cannot call wait on it. */
+
+static inline void
+fd_tpool_wait( fd_tpool_t const * tpool,
+               ulong              worker_idx ) {
+  int volatile * vstate = (int volatile *)&(fd_tpool_private_worker( tpool )[ worker_idx ]->state);
+  int            state;
+  for(;;) {
+    state = *vstate;
+    if( FD_LIKELY( state!=FD_TPOOL_WORKER_STATE_EXEC ) ) break;
+    FD_SPIN_PAUSE();
+  }
+}
+
+/* Assuming the tasks can be executed safely in any order and/or
+   concurrently, fd_tpool_exec_all_rrobin, fd_tpool_exec_all_block and
+   fd_tpool_exec_all_taskq are functionally equivalent to:
+
+     for( ulong l=l0; l<l1; l++ )
+       task( task_tpool,t0,t1, task_args,task_reduce,task_stride, task_l0,task_l1, l,l+1, t,t+1 );
+
+   where t above is in [t0,t1) and gives the tpool worker thread idx
+   assigned the particular task (and thread t0 is the thread that did
+   the dispatch).  The rrobin variant stripes individual tasks
+   deterministically over worker threads (e.g. worker thread t does
+   tasks (t-t0)+0*(t1-t0),(t-t0)+1*(t1-t0),(t-t0)+2*(t1-t0), ...  The
+   block variant blocks individual tasks over worker threads.
+   
+   The taskq variant requires FD_HAS_ATOMIC support and assigns tasks to
+   threads dynamically.  Practically, this is only useful if there are a
+   huge number of tasks to execute relative to the number of threads,
+   the tasks to execute have highly non-uniform sizes, the cost to
+   execute a task is much much greater than the cost of a single atomic
+   memory operation, and thread-core affinity doesn't impact the time to
+   execute a task ... conditions that, in total, are met far less
+   frequently than most developers expect.)
+
+   fd_tpool_exec_all_batch is functionally equivalent to:
+
+     for( ulong t=t0; t<t1; t++ ) {
+       ulong batch_task_l0;
+       ulong batch_task_l1;
+       FD_TPOOL_PARTITION( task_l0,task_l1,1, t-t0,t1-t0, batch_task_l0,batch_task_l1 );
+       task( task_tpool,t0,t1, task_args,task_reduce,task_stride, task_l0,task_l1, batch_task_l0,batch_task_l1, t,t+1 );
+     }
+
+   The batch assigned to a thread will be the same tasks as those
+   assigned to a thread by fd_tpool_exec_all_block.
```

Comment:
I did not understand the difference between this and `fd_tpool_exec_all_block` until reading the test. 
```suggestion
   The batch assigned to a thread will be the same tasks as those
   assigned to a thread by fd_tpool_exec_all_block.  The difference
   is that fd_tpool_exec_all_batch effectively merges all the calls to task
   that fd_tpool_exec_all_block would make in a block into a single call
   per thread.
```